pep-0xxx.rst

```diff
@@ -0,0 +1,1250 @@
+            from ``incoming`` and decrypt it, and write encrypted data to
+            ``outgoing``.
+
+            The buffer objects must be either file objects or objects that
+            implement the buffer protocol.
+            """
+
+
+Socket
+~~~~~~
+
+The socket-wrapper ABC will be defined by the ``TLSWrappedSocket`` ABC, which
+has the following definition::
+
+    class TLSWrappedSocket(metaclass=ABCMeta):
+        # The various socket methods all must be implemented. Their definitions
+        # have been elided from this class defintion in the PEP because they
+        # aren't instructive.
+        @abstractmethod
+        def do_handshake(self) -> None:
+            """
+            Performs the TLS handshake. Also performs certificate validation
+            and hostname verification.
+            """
+
+        @abstractmethod
+        def cipher(self) -> Optional[CipherSuite]:
+            """
+            Returns the CipherSuite entry for the cipher that has been
+            negotiated on the connection. If no connection has been negotiated,
+            returns ``None``.
+            """
+
+        @abstractmethod
+        def negotiated_protocol(self) -> Optional[Union[NextProtocol, bytes]]:
+            """
+            Returns the protocol that was selected during the TLS handshake.
+            This selection may have been made using ALPN, NPN, or some future
+            negotiation mechanism.
+
+            If the negotiated protocol is one of the protocols defined in the
+            ``NextProtocol`` enum, the value from that enum will be returned.
+            Otherwise, the raw bytestring of the negotiated protocol will be
+            returned.
+
+            If ``Context.set_inner_protocols()`` was not called, if the other
+            party does not support protocol negotiation, if this socket does
+            not support any of the peer's proposed protocols, or if the
+            handshake has not happened yet, ``None`` is returned.
+            """
+
+        @property
+        @abstractmethod
+        def context(self) -> Context:
+            """
+            The ``Context`` object this socket is tied to.
+            """
+
+        @abstractproperty
+        def negotiated_tls_version(self) -> Optional[TLSVersion]:
+            """
+            The version of TLS that has been negotiated on this connection.
+            """
+
+        @abstractmethod
+        def unwrap(self) -> socket.socket:
+            """
+            Cleanly terminate the TLS connection on this wrapped socket. Once
+            called, this ``TLSWrappedSocket`` can no longer be used to transmit
+            data. Returns the socket that was wrapped with TLS.
+            """
+
+
+Buffer
+~~~~~~
+
+The buffer-wrapper ABC will be defined by the ``TLSWrappedBuffer`` ABC, which
+has the following definition::
+
+    class TLSWrappedBuffer(metaclass=ABCMeta):
```

Comment:
@Lukasa - Recent changes to the Twisted TLS stack are actually quite illuminating in this regard :).

It's tempting to think of TLS as purely a layered byte-stream (and in many cases that abstraction is helpful) but in fact it has several slightly odd corners, particularly around handshaking.  If it were just a byte-stream, then interface 1 would be the preferable one, but interface 2 is more reflective of the fact that code which wants to be retrieving wire bytes to send may be calling `do_handshake()` first _or_ calling `bytes_from_wire(data)` first, depending on where it is in the state machine.

So I think I'm voting for 2.

Also, IMHO, "read" and "write" are _incredibly_ confusing names for these methods, despite their use being somewhat widespread.  I would much prefer `bytes_from_wire(data)` and `bytes_from_application(data)` or something similar, as well as a symmetrical `bytes_for_wire()` and `bytes_for_application()` for unspooling the buffer.

I will probably have more thoughts on this later, but is this sufficient to answer the question now?